research/conditional/moe_layers/token_choice.py

```diff
@@ -81,57 +87,84 @@ def forward(self, x: torch.Tensor):
+
+        n_selected_tokens = truncated_expert_mask.sum().item()
+
+        self.update_cache_for_logging(
+            "dropped_tokens_ratio",
+            ((n_tokens * self.routing_top_k) - n_selected_tokens)
+            / (n_tokens * self.routing_top_k),
+        )
 
         with measure_time(self, "calculate aux loss"):
-            position_in_expert_mask = position_in_expert.bool()
-            tokens_per_expert = position_in_expert_mask.sum(dim=0, dtype=gate_out.dtype)
+            tokens_per_expert = expert_mask.sum(dim=0, dtype=gate_out.dtype)
             load_balancing_loss = calculate_load_balancing_loss(
                 self.load_balancing_loss_weight,
                 gate_out,
                 tokens_per_expert,
                 use_einsum=self.use_einsum,
             )
 
-            if "load_balancing_losses" not in self.forward_pass_cache:
-                self.forward_pass_cache["load_balancing_losses"] = [load_balancing_loss]
-            else:
-                self.forward_pass_cache["load_balancing_losses"].append(
-                    load_balancing_loss
-                )
+        if "load_balancing_losses" not in self.forward_pass_cache:
+            self.forward_pass_cache["load_balancing_losses"] = [load_balancing_loss]
+        else:
+            self.forward_pass_cache["load_balancing_losses"].append(load_balancing_loss)
 
         self.update_cache_for_logging("gate_softmax_values", expert_gate)
         self.update_cache_for_logging("max_indices", expert_index)
         self.update_cache_for_logging("tokens_per_expert", tokens_per_expert)
         self.update_cache_for_logging("load_balancing_loss", load_balancing_loss)
-        # group tokens indices by expert it should be processed by
-        with measure_time(self, "experts_lists"):
-            indices_of_tokens_for_expert = [
-                single_expert_mask.nonzero(as_tuple=True)[0]
-                for single_expert_mask in expert_mask.transpose(0, 1)
-            ]
 
-        masked_expert_gate = gate_out * expert_mask
+        masked_expert_gate = gate_out * truncated_expert_mask
 
         # create empty input and  assign only tokens to be processed
         experts_input = torch.zeros(
             self.n_experts, capacity, self.dmodel, dtype=x.dtype, device=x.device
         )
-        with measure_time(self, "assign_tokens_to_input"):
-            for i, indices in enumerate(indices_of_tokens_for_expert):
-                experts_input[i, : len(indices)] = x[indices]
-
-        return experts_input, indices_of_tokens_for_expert, masked_expert_gate
+        if self.vectorize:
+            with measure_time(self, "assign_tokens_to_input"):
+                experts_input = x[
+                    top_tokens_per_expert_indices.T.reshape(
+                        (self.n_experts * capacity,)
+                    ),
+                    :,
+                ] * top_tokens_per_expert_values.T.reshape(
```

Comment:
Why do we multiply by top_tokens_per_expert_values here? It looks differently to how it worked before.

Reply:
Ok, now I get it.